stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/querychannels/internal/QueryChannelsStateLogicImpl.kt

```diff
@@ -0,0 +1,237 @@
+     * or loading more channels.
+     */
+    override fun loadingForCurrentRequest(): MutableStateFlow<Boolean> {
+        return mutableState._currentRequest.value?.isFirstPage?.let { isFirstPage ->
+            if (isFirstPage) mutableState._loading else mutableState._loadingMore
+        } ?: mutableState._loading
+    }
+
+    /**
+     * Add channels to state
+     *
+     * @param channels List<Channel>.
+     */
+    override fun addChannelsState(channels: List<Channel>) {
+        mutableState.queryChannelsSpec.cids += channels.map { it.cid }
+        val existingChannels = mutableState.rawChannels ?: emptyMap()
+        mutableState.rawChannels = existingChannels + channels.map { it.cid to it }
+        channels.forEach { channel ->
+            logicRegistry.channelState(channel.type, channel.id).updateDataFromChannel(
+                channel,
+                shouldRefreshMessages = false,
+                scrollUpdate = false
+            )
+        }
+    }
+
+    /**
+     * Remove channels to state.
+     */
+    override fun removeChannels(cidSet: Set<String>) {
+        val existingChannels = mutableState.rawChannels ?: return
+
+        mutableState.queryChannelsSpec.cids = mutableState.queryChannelsSpec.cids - cidSet
+        mutableState.rawChannels = existingChannels - cidSet
+    }
+
+    /**
+     * Initializes [QueryChannelsMutableState.rawChannels] with an empty map if it wasn't initialized yet.
+     * This might happen when we don't have any channels in the offline storage and API request fails.
+     */
+    override fun initializeChannelsIfNeeded() {
+        if (mutableState.rawChannels == null) {
+            mutableState.rawChannels = emptyMap()
+        }
+    }
+
+    /**
+     * Refreshes multiple channels in this query.
+     * Note that it retrieves the data from the current [ChannelState] object.
+     *
+     * @param cidList The channels to refresh.
+     */
+    override fun refreshChannels(cidList: Collection<String>) {
+        val existingChannels = mutableState.rawChannels
+        if (existingChannels == null) {
+            logger.w { "[refreshChannels] rejected (existingChannels is null)" }
+            return
+        }
+        mutableState.rawChannels = existingChannels + mutableState.queryChannelsSpec.cids
+            .intersect(cidList.toSet())
+            .map { cid -> cid.cidToTypeAndId() }
+            .filter { (channelType, channelId) ->
+                stateRegistry.isActiveChannel(
+                    channelType = channelType,
+                    channelId = channelId,
+                )
+            }
+            .associate { (channelType, channelId) ->
+                val cid = (channelType to channelId).toCid()
+                cid to stateRegistry.channel(
+                    channelType = channelType,
+                    channelId = channelId,
+                ).toChannel()
+            }
```

Comment:
This logic shouldn't be done here.
MutableStates should be udpated into the MutableState, and only expose "read-only states"